File: week 5/primitive_calculator/PrimitiveCalculator.java
import java.util.*;

public class PrimitiveCalculator {
    private static List<Integer> optimal_sequence(int n) {
        List<Integer> sequence = new ArrayList<Integer>();
        while (n >= 1) {
            sequence.add(n);
            if (n % 3 == 0) {
                n /= 3;
            } else if (n % 2 == 0) {
                n /= 2;
            } else {
                n -= 1;
            }
        }
        Collections.reverse(sequence);
        return sequence;
    }
	
	private static List<Integer> dp_sequence(int n) {
		List<Integer> sequence = new ArrayList<Integer>();
		int [] a = new int[n+1];
		for(int i =1; i< a.length; i++){
			a[i] = a[i-1]+1;
			if (i%2 == 0) 
				a[i] = Math.min(1+a[i/2], a[i]);
			if (i%3 == 0) 
				a[i] = Math.min(1+a[i/3], a[i]);

		}
		int i = n;
		while (i>1) {
			sequence.add(i);
			if (a[i-1] == a[i]-1)
				i = i-1;
			else if (i%2 == 0 && (a[i/2] == a[i]-1))
				i = i/2;
			else if (i%3 == 0 && (a[i/3] == a[i]-1))
				i = i/3;
		}
		sequence.add(1);
		Collections.reverse(sequence);
		return sequence;
	}

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        List<Integer> sequence = dp_sequence(n);
        System.out.println(sequence.size() - 1);
        for (Integer x : sequence) {
            System.out.print(x + " ");
        }
    }
}

File: week 5/primitive_calculator/primitive_calculator.py
def optimal_sequence(n):
    sequence = []
    while n >= 1:
        sequence.append(n)
        if n % 3 == 0:
            n = n // 3
        elif n % 2 == 0:
            n = n // 2
        else:
            n = n - 1
    return reversed(sequence)

def dp_sequence(n):
    sequence = []
    a = [0]*(n+1)
    for i in range(1,len(a)):
        a[i] = a[i-1] + 1
        if (i%2 == 0):
            a[i] = min(1+a[i/2], a[i])
        if (i%3 == 0):
            a[i] = min(1+a[i/3], a[i])
    while (n>1):
        sequence.append(n)
        if (a[n-1] == a[n]-1):
            n = n-1
        elif (n%2 == 0 and (a[n/2] == a[n]-1)):
            n = n/2		
        elif (n%3 == 0 and (a[n/3] == a[n]-1)):
            n = n/3
    sequence.append(1)
    return reversed(sequence)

n = int(input())
sequence = list(dp_sequence(n))
print(len(sequence) - 1)
for x in sequence:
    print str(x) + ' ',